### PR TITLE
Add -internalize-at-link flag to allow dead stripping and VFE across modules at link time

### PR DESCRIPTION
**Pull request by kubamracek**
The first commit in the PR is being reviewed at <https://github.com/apple/swift/pull/39128>. Suggestion for reviewers is to only look at the second commit here.

- Under -internalize-at-link, stop unconditionally marking all globals as used.
- Under -internalize-at-link, restrict visibility of vtables to linkage unit.
- Emit virtual method thunks for cross-module vcalls when VFE is enabled.
- Use thunks for vcalls across modules when VFE is enabled.
- Adjust TBDGen to account for virtual method thunks when VFE is enabled.
- Add an end-to-end test case for cross-module VFE.

### CODE REVIEW

File: lib/IRGen/GenMeta.cpp [hunk 1]
```diff
@@ -1782,7 +1791,13 @@ namespace {
       // method for external clients.
       
       // Emit method dispatch thunk.
-      if (hasPublicVisibility(fn.getLinkage(NotForDefinition))) {
+      bool shouldEmitDispatchThunk = false;
```
Comment (eeckstein) [inline lib/IRGen/GenMeta.cpp:1794]: Can this logic, which is the same as the previous change, be extracted into a utility function?

Comment (kubamracek) [reply lib/IRGen/GenMeta.cpp]: It wasn't entirely identical, so I've instead simplified it into just an or-d expression inside the if.

**Review by eeckstein — APPROVED** [review]
LGTM

**Comment by kubamracek** [comment]
@swift-ci please test Linux platform

**Pull request merged**